### PR TITLE
Fix #58598 X-ID-Token header missing on Loki Datasource

### LINKED ISSUE #58598: X-ID-Token header still missing on Loki Datasource
**What happened**:
[#57878](https://github.com/grafana/grafana/pull/57878) added the `X-ID-Token` header to the Loki datasource but when enabling the _Forward OAuth Identity_, the header is still missing. The header is passed correctly to the Prometheus datasource.

**What you expected to happen**:
See the `X-ID-Token` in the headers passed to Loki.

**How to reproduce it (as minimally and precisely as possible)**:
1. Configure Grafana with OAuth-based user authentication (generic_oauth with Auth0 in my case).
2. Configure a Loki-based datasource, where option _Forward OAuth Identity_ is enabled.
3. Check whether `X-ID-Token` is passed to the Loki backend. Which it is not, in my case.

**Anything else we need to know?**:

According to the `net/http.Request` documentation the header names are case-insensitive and are parsed using `CanonicalHeaderKey`. This results in the `X-ID-Token` being converted to `X-Id-Token` when added to the request Headers here https://github.com/grafana/grafana/blob/release-9.2.4/pkg/api/plugin_resource.go#L86 and later not matching the name here https://github.com/grafana/grafana/blob/release-9.2.4/pkg/tsdb/loki/loki.go#L133

Happy to submit a PR to fix this issue but would need some guidance. Few initial ideas:
1. Add an extra `if` block to get the header using `X-Id-Token` (horrible but fixed the issue. Tested locally)
2. Rename all the references of `X-ID-Token` to `X-Id-Token`. Not sure of the side effects this would introduce.
3. Any other suggestions?

**Environment**:
- Grafana version: 9.2.4 (reproduced on main branch)
- Data source type & version: Loki 2.6.1
- OS Grafana is installed on: MacOS + Official Docker Alpine image
- User OS & Browser: MacOS, Firefox
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by yvigara**
**What is this feature?**
Refactor the Loki datasource to extract headers using `CanonicalMIMEHeaderKey` formating.

**Why do we need this feature?**
[57878](https://github.com/grafana/grafana/pull/57878) added the X-ID-Token header to the Loki datasource but when enabling the Forward OAuth Identity, the header is still missing. The header is passed correctly to the Prometheus datasource.

**Who is this feature for?**
Grafana administrators and datasource operators configuring authentication for Grafana datasources and their backends.

**Which issue(s) does this PR fix?**:
Fixes #58598

### CODE REVIEW

File: pkg/tsdb/loki/loki.go [hunk 1]
```diff
@@ -122,16 +123,23 @@ func (s *Service) CallResource(ctx context.Context, req *backend.CallResourceReq
...
+	for k, values := range headers {
+		k = textproto.CanonicalMIMEHeaderKey(k)
+		firstValue := arrayHeaderFirstValue(values)
 
-	if idToken := arrayHeaderFirstValue(headers["X-ID-Token"]); idToken != "" {
-		data["X-ID-Token"] = idToken
+		if firstValue == "" {
+			continue
+		}
+		switch k {
```
Comment (gabor) [inline pkg/tsdb/loki/loki.go:133]: i don't think you need the `continue` statements here. there is no more code after the `switch` block, and in go only the matches `case` is executed as far as i know ( https://go.dev/tour/flowcontrol/9 )
or maybe i'm missing something 😄

**Review by gabor — DISMISSED** [review]
thanks for the PR! i haven't tested it yet, just did an initial read of the `diff`. i left a comment.

File: pkg/tsdb/loki/loki.go (hunk 1 above)
Comment (yvigara) [reply pkg/tsdb/loki/loki.go:133]: Yes you are right. I didn't put too much thought after replicating what I saw on this PR: https://github.com/grafana/grafana/pull/58506
Should be fixed now.

**Comment by yvigara** [comment]
> also, it would be great to add this to a unit test somehow... could we add it maybe to the test at https://github.com/grafana/grafana/blob/main/pkg/tests/api/loki/loki_test.go ?

Looks like there's already some Unit test here: https://github.com/grafana/grafana/blob/main/pkg/tsdb/loki/auth_test.go
I'll spend some time later today to add something to the integration test you refered to.

**Comment by yvigara** [comment]
@gabor I've added unit tests for the `getAuthHeadersForCallResource` method. The https://github.com/grafana/grafana/blob/main/pkg/tsdb/loki/auth_test.go is covering authentication test cases.
I didn't find an easy way to test OAuth Forwarding config in the integration tests https://github.com/grafana/grafana/blob/main/pkg/tests/api/loki/loki_test.go. Looks like this would require creating some boilerplate to setup the OAuth authentication first

**Comment by svennergr** [comment]
Thanks for your contribution @yvigara. We recently did other changes regarding those headers in https://github.com/grafana/grafana/pull/59059. Thus this PR needs a merge from the current `main` branch. 

Are you able to merge those changes into this PR?

Thanks!

File: pkg/tsdb/loki/loki.go [hunk 2]
```diff
@@ -122,22 +123,25 @@ func (s *Service) CallResource(ctx context.Context, req *backend.CallResourceReq
...
+		case "Authorization":
+			data["Authorization"] = firstValue
+		case "X-Id-Token":
+			data["X-ID-Token"] = firstValue
+		case "Cookie":
+			data["Cookie"] = firstValue
+		}
+    case "Accept-Encoding":
+			data["Accept-Encoding"] = firstValue
+		}
```
Comment (svennergr) [inline pkg/tsdb/loki/loki.go:143]: I think this needs to move up one line, right?

**Review by svennergr — CHANGES_REQUESTED** [review]

File: pkg/tsdb/loki/loki.go (hunk 2 above)
Comment (yvigara) [reply pkg/tsdb/loki/loki.go]: Yes, sorry I missed that when merging and then got distracted.
I also fixed the test that was failing after the merge and added the `Accept-Encoding` header to the test case

File: pkg/tsdb/loki/loki.go [hunk 3]
```diff
@@ -122,22 +123,24 @@ func (s *Service) CallResource(ctx context.Context, req *backend.CallResourceReq
...
-	if encType := arrayHeaderFirstValue(headers["Accept-Encoding"]); encType != "" {
-		data["Accept-Encoding"] = encType
+		if firstValue == "" {
+			continue
+		}
+		switch k {
+		case "Authorization":
+			data["Authorization"] = firstValue
+		case "X-Id-Token":
+			data["X-ID-Token"] = firstValue
```
Comment (svennergr) [inline pkg/tsdb/loki/loki.go:137]: Just a nit, but I think we can go ahead and stick to using `X-Id-Token` here.

**Review by svennergr — APPROVED** [review]
I just verified that those changes work. Great job and thanks again for your contribution!

~~I left a comment, but feel free to merge without the change as it is just a nit.~~

Disregard the above, I'll merge it without the change.

**Pull request merged**